kernel/kernel-api/src/main/java/io/delta/kernel/internal/DeltaHistoryManager.java

```diff
@@ -0,0 +1,249 @@
+            Optional<Long> lastCompleteCheckpoint = Optional.empty();
+
+            // Iterate through the log files - this will be in order starting from the lowest
+            // version. Checkpoint files come before deltas, so when we see a checkpoint, we
+            // remember it and return it once we detect that we've seen a smaller or equal delta
+            // version.
+            while (files.hasNext()) {
+                String nextFilePath = files.next().getPath();
+                if (FileNames.isCommitFile(getName(nextFilePath))) {
+                    long version = FileNames.deltaVersion(nextFilePath);
+                    if (version == 0L) {
+                        return version;
+                    }
+                    smallestDeltaVersion = Math.min(version, smallestDeltaVersion);
+
+                    // Note that we also check this condition at the end of the function - we check
+                    // it here too to try and avoid more file listing when it's unnecessary.
+                    if (lastCompleteCheckpoint.isPresent() &&
+                        lastCompleteCheckpoint.get() >= smallestDeltaVersion ) {
+                        return lastCompleteCheckpoint.get();
+                    }
+                } else if (FileNames.isCheckpointFile(nextFilePath)) {
+                    long checkpointVersion = FileNames.checkpointVersion(nextFilePath);
+                    CheckpointInstance checkpointInstance = new CheckpointInstance(nextFilePath);
+                    if (!checkpointInstance.numParts.isPresent()) {
+                        lastCompleteCheckpoint = Optional.of(checkpointVersion);
+                    } else {
+                        // if we have a multi-part checkpoint, we need to check that all parts exist
+                        int numParts = checkpointInstance.numParts.orElse(1);
+                        int preCount = checkpointMap.getOrDefault(
+                            new Tuple2<>(checkpointVersion, numParts), 0);
+                        if (numParts == preCount + 1) {
+                            lastCompleteCheckpoint = Optional.of(checkpointVersion);
+                        }
+                        checkpointMap.put(new Tuple2<>(checkpointVersion, numParts), preCount + 1);
+                    }
+                }
+            }
+
+            if (lastCompleteCheckpoint.isPresent() &&
+                lastCompleteCheckpoint.get() >= smallestDeltaVersion) {
+                return lastCompleteCheckpoint.get();
+            } else if (smallestDeltaVersion < Long.MAX_VALUE) {
+                // This is a corrupt table where 000.json does not exist and there are no complete
+                // checkpoints OR the earliest complete checkpoint does not have a corresponding
+                // commit file (but there are other later commit files present)
+                throw new RuntimeException(
+                    String.format("No recreatable commits found at %s", logPath));
```

Comment:
We may need to create an InvalidTableStateException to capture all these runtime exceptions.

Reply:
Agreed let's add this to the exception discussion